rfcs/037-index-query-composition.md

```diff
@@ -0,0 +1,115 @@
+        - user: an optional user that executed the query (should be Amundsen user ID)
+    - Relations:
+        - Table > Query
+        - User > Query
+2. Query Execution
+    - Attributes:
+        - query: the query model defined above
+        - start time: The start time of the window that the aggregation represents. Query executions can be dropped from the database when start time > X in order to update the most commonly used joins and wheres.
+        - execution_count: the total # of times the query was executed in the window
+        - window_duration: a string representation of the window length. Amundsen will not enforce non-overlapping windows, users should schedule a job to create query executions on defined cadence.
+    - Relations:
+        - Query >> Query Execution
+3. Join
+    - Attributes:
+        - left_table: Table Metadata for the left table in a join
+        - right_table: Table Metadata for the right table in a join
+        - left_column: Column Metadata for the left column in a join
+        - right_column: Column Metadata for the right column in a join
+        - join_type: a string representing the join type, could be `inner join`, `left outer join`, etc
+        - join_operator: a string representing the operator, could be `=`, `>`, `<=`, etc.
+        - join_sql: the sql code for this join (e.g. `table join table2 on table.a = table2.b`)
+        - query: The query model defined above
+    - Relations:
+        - Column >> Join
+        - Query >> Join
+4. Where
+    - Attributes:
+        - tables: a list of Table Metadata representing each table used in the query. The columns in the table metadata should be used in the where clause
+        - where_clause: a string representation of the SQL where (e.g. `field < 4` or `val is null`)
+        - query_metadata: The query model defined above
+    - Relations:
+        - Column >> Where
+        - Query >> Where
+
+## Drawbacks
+
+Parsing query logs can be a large effort, therefore this approach will be designed in a way such that implementers will be able to send a raw query into Amundsen without the other components (executions, joins and wheres) but will loose the ability for the features built on top of those data assets such as ordering by most common query, showing the most common filters, etc. However, Amundsen is not intended to hold a copy of each query each time it is executed and implementers should take some effort to dedupe queries.
+
+## Alternatives
+
+- Do not index queries and query components
+- Only index queries, not components of queries
+
+## Prior art
+
+[PR 27](https://github.com/amundsen-io/rfcs/pull/27) had been opened to add queries as first class objects. This PR did a good job to raise awareness of the need for having queries as first-class objects but this PR expands on that in a few areas:
```

Comment:
are you talking about making query as entity like table? what is the benefit compared to showing the top used query in table tab / dashboard tab?

Reply:
I haven't seen any other solutions has been put query as 1st class entity. The user benefit is unclear compared to putting query under table<->dashboard which are more user-phasing( to me query node is similar to column node that belong to table/dashboard. )

Reply:
Sorry, I used unclear language and gave the wrong impression.

The intention is to make queries a top level concept at the data level, but not in the UI. Queries cannot be conceptually under tables (like a column can), because a single query touches multiple tables. However, we certainly would display them that way (on the table detail page, as you mentioned)

We will revise the RFC to make this clearer.

Reply:
At a high level what we're trying to do is to help users discover _how to use their data_ at the same time that they discover the location and meaning of their data. 

As to why create these as first-class objects in the database, there are some questions that are difficult to answer without it:

- What are the most common filters used _now_, this is especially important as data evolves over time (supported by `(Where) -> (Query) -> (Query Execution)` relationship)


In addition, there is the added benefit of referencing a single query object, a single join object, etc. instead of replicating that for each table / dashboard that it is associated to.